runtime/src/main/java/org/corfudb/runtime/object/VersionLockedObject.java

```diff
@@ -672,25 +674,32 @@ protected void syncStreamUnsafe(ISMRStream stream, long timestamp) {
         log.trace("Sync[{}] {}", this, (timestamp == Address.OPTIMISTIC)
                 ? "Optimistic" : "to " + timestamp);
         long syncTo = (timestamp == Address.OPTIMISTIC) ? Address.MAX : timestamp;
-        stream.streamUpTo(syncTo)
-                .forEachOrdered(entry -> {
-                    try {
-                        Object res = applyUpdateUnsafe(entry, timestamp);
-                        if (timestamp == Address.OPTIMISTIC) {
+        Timer timer = VloMetricsHelper.metrics.timer(VloMetricsHelper.VLO_SYNC_STREAM + stream.getID());
+        Histogram histogram = VloMetricsHelper.metrics.histogram(VloMetricsHelper.VLO_SYNC_COUNT + stream.getID());
+        AtomicLong counter = new AtomicLong();
+        try (Timer.Context context = MetricsUtils.getConditionalContext(timer)) {
+            stream.streamUpTo(syncTo)
+                    .forEachOrdered(entry -> {
+                        try {
+                            counter.getAndIncrement();
+                            Object res = applyUpdateUnsafe(entry, timestamp);
+                            if (timestamp == Address.OPTIMISTIC) {
+                                entry.setUpcallResult(res);
+                            } else if (pendingUpcalls.contains(entry.getGlobalAddress())) {
+                                log.debug("Sync[{}] Upcall Result {}",
+                                        this, entry.getGlobalAddress());
+                                upcallResults.put(entry.getGlobalAddress(), res == null
+                                        ? NullValue.NULL_VALUE : res);
+                                pendingUpcalls.remove(entry.getGlobalAddress());
+                            }
                             entry.setUpcallResult(res);
-                        } else if (pendingUpcalls.contains(entry.getGlobalAddress())) {
-                            log.debug("Sync[{}] Upcall Result {}",
-                                    this, entry.getGlobalAddress());
-                            upcallResults.put(entry.getGlobalAddress(), res == null
-                                    ? NullValue.NULL_VALUE : res);
-                            pendingUpcalls.remove(entry.getGlobalAddress());
+                        } catch (Exception e) {
+                            log.error("Sync[{}] Error: Couldn't execute upcall due to {}", this, e);
+                            throw new UnrecoverableCorfuError(e);
                         }
-                        entry.setUpcallResult(res);
-                    } catch (Exception e) {
-                        log.error("Sync[{}] Error: Couldn't execute upcall due to {}", this, e);
-                        throw new UnrecoverableCorfuError(e);
-                    }
-                });
+                    });
+        }
+        histogram.update(counter.longValue());
```

Comment:
This might produce a bimodal distribution. Since the optimistic sync is much faster (or should be) than the materialized stream sync. 

Is it necessary to split them?

Reply:
Only measure the regular stream now.